### PR TITLE
[Txt2Img] CPU support + TorchScript and Intel® Extension for PyTorch* optimizations

### PR DESCRIPTION
**Pull request by aalbersk**
## Summary
Proposing following changes that enable CPU support for Stable Diffusion and add TorchScript and [Intel® Extension for PyTorch*(IPEX)](https://github.com/intel/intel-extension-for-pytorch) optimizations to improve the performance on Intel® CPU hardware.
The optimizations provide about 3x the performance comparing with currently implemented eager execution for Intel Xeon platform 3rd Generation(IceLake). These optimizations also enable BF-16 for Intel Xeon 4th Generation(Sapphire Rapids), which optimizes the throughput by another 3x.

 

## Changes
The code mainly implements TorchScript+IPEX optimizations in scripts/txt2img.py, but also adds support for CPU in samplers and provide bfloat16 support. The provided configs disable gradient checkpointing and add support for bfloat16 if needed.

### CODE REVIEW

**Comment by brycedrennan** [comment]
How much does this improve performance?

**Comment by aalbersk** [comment]
On AWS instance _c6i.16xlarge (Intel(R) Xeon(R) Platinum 8375C)_ the performance looks as follows:
|Batch size|Model name|Optimizations|Avg latency [in secs]|Boost in comparison with clean pytorch only CPU support|
|:----|:----|:----|:----|:----|
|1|SD2-v (768px)|clean, no ipex, only CPU support added|385.007|N/A|
|1|SD2-v (768px)|ipex + torchscript with ipex envs|110.356|349%|
|4|SD2-v (768px)|clean, no ipex, only CPU support added|1496.812|N/A|
|4|SD2-v (768px)|ipex + torchscript with ipex envs|460.320|325%|
|1|SD2-base (512px)|clean, no ipex, only CPU support added|112.521|N/A|
|1|SD2-base (512px)|ipex + torchscript with ipex envs|37.998|296%|
|4|SD2-base (512px)|clean, no ipex, only CPU support added|412.438|N/A|
|4|SD2-base (512px)|ipex + torchscript with ipex envs|145.346|284%|

BF-16 reduces the latency by another 3x with Intel 4th Gen Xeon Platform.

**Comment by UmaisZahid** [comment]
It might be useful to allow "xpu" as a possible device option so that Intel's new discrete GPU line works also. I think currently this would raise a ValueError. The discrete GPUs would like benefit from the Intel Extension optimizations given their XMX cores.

**Comment by aalbersk** [comment]
> It might be useful to allow "xpu" as a possible device option so that Intel's new discrete GPU line works also. I think currently this would raise a ValueError. The discrete GPUs would like benefit from the Intel Extension optimizations given their XMX cores.

We haven't tested this pipeline on XPU yet, but it's a great suggestion. We'll consider testing it in the future.

**Review by dmarx — APPROVED** [review]

**Comment by camenduru** [comment]
```py
Installing requirements for CodeFormer
Installing requirements for Web UI
Installing Deforum requirement: av
Installing Deforum requirement: pims
Installing Deforum requirement: imageio_ffmpeg


Launching Web UI with arguments: --share --xformers --enable-insecure-extension-access --gradio-queue
2023-02-11 20:33:38.960007: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 AVX512F FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2023-02-11 20:33:41.763634: W tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libnvinfer.so.7'; dlerror: libnvinfer.so.7: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /usr/local/lib/python3.8/dist-packages/cv2/../../lib64:/usr/lib64-nvidia
2023-02-11 20:33:41.764023: W tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libnvinfer_plugin.so.7'; dlerror: libnvinfer_plugin.so.7: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /usr/local/lib/python3.8/dist-packages/cv2/../../lib64:/usr/lib64-nvidia
2023-02-11 20:33:41.764049: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Cannot dlopen some TensorRT libraries. If you would like to use Nvidia GPU with TensorRT, please make sure the missing libraries mentioned above are installed properly.
[AddNet] Updating model hashes...
0it [00:00, ?it/s]
[AddNet] Updating model hashes...
0it [00:00, ?it/s]
```
